clang/docs/LanguageExtensions.rst

```diff
@@ -3774,6 +3774,74 @@ type-generic alternative to the ``__builtin_clz{,l,ll}`` (respectively
 ``__builtin_ctz{,l,ll}``) builtins, with support for other integer types, such
 as ``unsigned __int128`` and C23 ``unsigned _BitInt(N)``.
 
+``__builtin_counted_by_ref``
+----------------------------
+
+``__builtin_counted_by_ref`` returns a pointer to the count field from the
+``counted_by`` attribute.
+
+The argument must be a flexible array member. If the argument isn't a flexible
+array member or doesn't have the ``counted_by`` attribute, the builtin returns
+``(void *)0``.
```

Comment:
I don't think that would work. The main way this builtin will be used (at least initially) is in a macro, where we're not sure if the flexible array member has the `counted_by` attribute. Even if we had `__builtin_has_attribute`, it would still cause problems:

```c.
if (__builtin_has_attribute(p->array, counted_by))
  *__builtin_counted_by_ref(p->array) = 42;  /* <-- an error is produced */
```

Reply:
Ah, I was thinking the only valid use of that macro would be when the array has a `counted_by` attribute on it and so if you used the macro on something that's uncounted, you would get a diagnostic telling you to go fix the declaration so it has the information.